web/src/util.ts

```diff
@@ -302,45 +319,74 @@ export function clean_user_content_links(html: string): string {
-            // the browser's built in handling of the title attribute.
-            const title = elt.getAttribute("title");
-            if (title !== null) {
-                elt.setAttribute("aria-label", title);
-                elt.removeAttribute("title");
+export function initialize(): void {
+    $("#main_div").on(
+        "mouseenter",
+        ".rendered_markdown a",
+        (e: JQuery.TriggeredEvent<HTMLElement, undefined, HTMLElement, HTMLElement>) => {
+            let elt: HTMLElement = e.target;
+            if (elt.parentElement && elt.nodeName === "IMG") {
+                elt = elt.parentElement;
             }
-        } else {
-            // For non-image user uploads, the following block ensures that the title
-            // attribute always displays the filename as a security measure.
-            let title: string;
-            let legacy_title: string;
+            const href = elt.getAttribute("href");
+            const title = elt.getAttribute("title");
+            const aria_label = elt.getAttribute("aria-label");
+            const parentElement = elt.parentElement;
             if (
-                url.origin === window.location.origin &&
-                url.pathname.startsWith("/user_uploads/")
+                !href ||
+                (title && !parentElement?.classList.contains("message_inline_image")) ||
+                aria_label ||
+                elt.classList.contains("previewable")
             ) {
-                // We add the word "download" to make clear what will
-                // happen when clicking the file.  This is particularly
-                // important in the desktop app, where hovering a URL does
-                // not display the URL like it does in the web app.
-                title = legacy_title = $t(
-                    {defaultMessage: "Download {filename}"},
-                    {filename: url.pathname.slice(url.pathname.lastIndexOf("/") + 1)},
-                );
+                return;
+            }
+            if (is_url_previewable(href)) {
+                elt.classList.add("previewable");
+                return;
+            }
+            const is_inline_image =
+                elt.parentElement && elt.parentElement.classList.contains("message_inline_image");
+            if (is_inline_image) {
+                // For inline images we want to handle the tooltips explicitly, and disable
+                // the browser's built in handling of the title attribute.
+                const title = elt.getAttribute("title");
+                if (title !== null) {
+                    elt.setAttribute("aria-label", title);
+                    elt.removeAttribute("title");
+                }
             } else {
-                title = url.toString();
-                legacy_title = href;
+                // For non-image user uploads, the following block ensures that the title
+                // attribute always displays the filename as a security measure.
```

Comment:
Can you comment on the security implications of moving this out of `clean_user_content_links`? Is the idea that we're no longer using `title` attributes or something?

Reply:
I can't think of any security concerns here as we are just moving the setting title part out of the `clean_user_content_links`, as we have to decide whether to show a title or not based on whether URL is previewable or not.